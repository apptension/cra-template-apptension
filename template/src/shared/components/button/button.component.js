import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { empty } from 'ramda';

import { Container } from './button.styles';
import { BUTTON_TYPES, BUTTON_TYPE_PRIMARY } from './button.constants';

export const ButtonComponent = ({ children, className, disabled, mode, onClick, ...other }) => (
  <ThemeProvider theme={{ mode, disabled }}>
    <Container onClick={onClick} className={className} disabled={disabled} {...other}>
      {children}
    </Container>
  </ThemeProvider>
);

ButtonComponent.propTypes = {
  children: PropTypes.any,
  className: PropTypes.any,
  disabled: PropTypes.bool,
  mode: PropTypes.oneOf(BUTTON_TYPES),
  onClick: PropTypes.func,
};

ButtonComponent.defaultProps = {
  mode: BUTTON_TYPE_PRIMARY,
  onClick: empty,
};
