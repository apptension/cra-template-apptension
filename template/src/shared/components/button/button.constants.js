export const BUTTON_TYPE_PRIMARY = 'primary';
export const BUTTON_TYPE_SECONDARY = 'secondary';

export const BUTTON_TYPES = [BUTTON_TYPE_PRIMARY, BUTTON_TYPE_SECONDARY];
