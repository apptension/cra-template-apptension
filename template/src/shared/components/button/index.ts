export { ButtonComponent as Button } from './button.component';
