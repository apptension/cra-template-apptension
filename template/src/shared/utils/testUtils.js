import React from 'react';
import PropTypes from 'prop-types';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Route, Router } from 'react-router';
import { IntlProvider } from 'react-intl';
import { DEFAULT_LOCALE, translationMessages } from '../../i18n';
import { store as fixturesStore } from '../../../fixtures/store';

export * from '@testing-library/react';

export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = '/') => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const pushSpy = jest.spyOn(history, 'push');
  return {
    history,
    pushSpy,
  };
};

export const ProvidersWrapper = ({ children, context = {} }) => {
  const { router = {}, store = fixturesStore, messages } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  return (
    <Router history={routerHistory}>
      <HelmetProvider>
        <IntlProvider {...intlProviderMockProps}>
          <Provider store={createStore(identity, store)}>
            <Route path={routePath}>{children}</Route>
          </Provider>
        </IntlProvider>
      </HelmetProvider>
    </Router>
  );
};

ProvidersWrapper.propTypes = {
  context: PropTypes.shape({
    router: PropTypes.shape({
      url: PropTypes.string,
      routePath: PropTypes.string,
      history: PropTypes.object,
    }),
    store: PropTypes.object,
    messages: PropTypes.object,
  }),
};

export const makeContextRenderer = component => (props = {}, context = {}) =>
  render(component(props), {
    wrapper: ({ children }) => <ProvidersWrapper context={context}>{children}</ProvidersWrapper>,
  });

export const makePropsRenderer = component => (props = {}) => render(component(props));
