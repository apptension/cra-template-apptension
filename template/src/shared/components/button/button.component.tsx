import React, { ReactNode, MouseEventHandler, FC } from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { empty } from 'ramda';

import { Container } from './button.styles';
import { BUTTON_TYPES, BUTTON_TYPE_PRIMARY } from './button.constants';

interface ButtonComponentProps {
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  mode?: string;
  onClick?: MouseEventHandler;
}

export const ButtonComponent: FC<ButtonComponentProps> = ({
  children,
  className,
  disabled,
  mode,
  onClick,
  ...other
}) => (
  <ThemeProvider theme={{ mode, disabled }}>
    <Container onClick={onClick} className={className} disabled={disabled} {...other}>
      {children}
    </Container>
  </ThemeProvider>
);

ButtonComponent.propTypes = {
  children: PropTypes.any,
  className: PropTypes.any,
  disabled: PropTypes.bool,
  mode: PropTypes.oneOf(BUTTON_TYPES),
  onClick: PropTypes.func,
};

ButtonComponent.defaultProps = {
  mode: BUTTON_TYPE_PRIMARY,
  onClick: empty,
};
