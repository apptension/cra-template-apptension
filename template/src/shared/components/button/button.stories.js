import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';

import { ButtonComponent } from './button.component';
import { BUTTON_TYPE_SECONDARY, BUTTON_TYPES, BUTTON_TYPE_PRIMARY } from './button.constants';

const knobs = {
  disabled: defaultValue => boolean('disabled', defaultValue),
  mode: defaultValue => select('mode', BUTTON_TYPES, defaultValue),
};

const renderComponent = (props = {}) => {
  const defaultProps = {
    children: 'Press me',
    onClick: action('Clicked me'),
    disabled: knobs.disabled(false),
    mode: knobs.mode(BUTTON_TYPE_PRIMARY),
  };

  return <ButtonComponent {...defaultProps} {...props} />;
};

const stories = storiesOf('Shared|Button', module).addDecorator(withKnobs);

stories.add('Primary', () => renderComponent());
stories.add('Primary - Disabled', () =>
  renderComponent({
    disabled: knobs.disabled(true),
  })
);
stories.add('Secondary', () =>
  renderComponent({
    mode: knobs.mode(BUTTON_TYPE_SECONDARY),
  })
);
stories.add('Secondary - Disabled', () =>
  renderComponent({
    mode: knobs.mode(BUTTON_TYPE_SECONDARY),
    disabled: knobs.disabled(true),
  })
);
