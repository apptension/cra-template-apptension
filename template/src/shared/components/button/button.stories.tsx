import React from 'react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';

import { ButtonComponent } from './button.component';
import { BUTTON_TYPE_SECONDARY, BUTTON_TYPES, BUTTON_TYPE_PRIMARY } from './button.constants';

const knobs = {
  disabled: (defaultValue: boolean) => boolean('disabled', defaultValue),
  mode: (defaultValue: string) => select('mode', BUTTON_TYPES, defaultValue),
};

const renderComponent = (props = {}) => {
  const defaultProps = {
    children: 'Press me',
    onClick: action('Clicked me'),
    disabled: knobs.disabled(false),
    mode: knobs.mode(BUTTON_TYPE_PRIMARY),
  };

  return <ButtonComponent {...defaultProps} {...props} />;
};

export default {
  title: 'Shared|Button',
  decorators: [withKnobs],
};

export const Primary = () => renderComponent();
export const PrimaryDisabled = () =>
  renderComponent({
    disabled: knobs.disabled(true),
  });

export const Secondary = () =>
  renderComponent({
    mode: knobs.mode(BUTTON_TYPE_SECONDARY),
  });

export const SecondaryDisabled = () =>
  renderComponent({
    mode: knobs.mode(BUTTON_TYPE_SECONDARY),
    disabled: knobs.disabled(true),
  });
