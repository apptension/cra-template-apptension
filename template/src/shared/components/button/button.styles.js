import styled, { css } from 'styled-components';
import theme from 'styled-theming';
import { BUTTON_TYPE_PRIMARY, BUTTON_TYPE_SECONDARY } from './button.constants';
import { styleWhenTrue, themeBorder, themeColor, themeSize } from '../../../theme/getters';
import { Color, Size } from '../../../theme/theme.constants';

const primaryButtonStyle = css`
  background: ${themeColor(Color.PRIMARY)};
`;

const secondaryButtonStyle = css`
  background: ${themeColor(Color.SECONDARY)};
`;

const disabledButtonStyle = css`
  background: ${themeColor(Color.DISABLED)};
  color: ${theme('mode', {
    [BUTTON_TYPE_PRIMARY]: themeColor(Color.BLACK),
    [BUTTON_TYPE_SECONDARY]: themeColor(Color.SECONDARY),
  })};
`;

export const Container = styled.button`
  padding: ${themeSize(Size.CONTENT_VERTICAL_PADDING)}px ${themeSize(Size.CONTENT_HORIZONTAL_PADDING)}px;
  border: ${themeBorder('regular')};
  ${theme('mode', {
    [BUTTON_TYPE_PRIMARY]: primaryButtonStyle,
    [BUTTON_TYPE_SECONDARY]: secondaryButtonStyle,
  })};
  ${styleWhenTrue('disabled', disabledButtonStyle)};
`;
