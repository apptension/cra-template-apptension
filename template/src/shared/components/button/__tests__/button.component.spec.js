import React from 'react';
import { makePropsRenderer, fireEvent, screen } from 'utils/testUtils';
import { Button } from '../button.component';

describe('Button: Component', () => {
  const defaultProps = {};

  const component = props => <Button {...defaultProps} {...props} />;
  const render = makePropsRenderer(component);

  it('should render correctly', () => {
    const { container } = render();
    expect(container).toMatchSnapshot();
  });

  it('should call onClick prop when clicked', () => {
    const onClick = jest.fn();
    render({ onClick });
    fireEvent.click(screen.getByRole('button'));
    expect(onClick).toHaveBeenCalled();
  });

  it('should pass native HTML props directly to the button element', () => {
    render({ type: 'submit' });
    expect(screen.getByRole('button')).toHaveAttribute('type', 'submit');
  });
});
