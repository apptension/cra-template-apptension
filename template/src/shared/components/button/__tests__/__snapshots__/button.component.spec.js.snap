// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button: Component should render correctly 1`] = `
.c0 {
  padding: px px;
}

<div>
  <button
    class="c0"
  />
</div>
`;
